plugins/experimental/ja3_fingerprint/ja3_fingerprint.cc

```diff
@@ -0,0 +1,489 @@
+    if (unit != 2 || GREASE_table.find(tmp) == GREASE_table.end()) {
+      if (!first) {
+        result += '-';
+      }
+      first = false;
+      result += std::to_string(tmp);
+    }
+  }
+  return 0;
+}
+
+char *
+getIP(sockaddr const *s_sockaddr, char res[INET6_ADDRSTRLEN])
+{
+  res[0] = '\0';
+
+  if (s_sockaddr == nullptr) {
+    return nullptr;
+  }
+
+  switch (s_sockaddr->sa_family) {
+  case AF_INET: {
+    const struct sockaddr_in *s_sockaddr_in = reinterpret_cast<const struct sockaddr_in *>(s_sockaddr);
+    inet_ntop(AF_INET, &s_sockaddr_in->sin_addr, res, INET_ADDRSTRLEN);
+  } break;
+  case AF_INET6: {
+    const struct sockaddr_in6 *s_sockaddr_in6 = reinterpret_cast<const struct sockaddr_in6 *>(s_sockaddr);
+    inet_ntop(AF_INET6, &s_sockaddr_in6->sin6_addr, res, INET6_ADDRSTRLEN);
+  } break;
+  default:
+    return nullptr;
+  }
+
+  return res[0] ? res : nullptr;
+}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+// Parsing clientHello to get ja3 string
+// No error checking or handling because this should be called after openSSL has done all checks and
+// returned successfully
+static std::string
+custom_get_ja3(SSL *s)
+{
+  TSDebug(PLUGIN_NAME, "Entering custom_get_ja3()...");
+  std::string ja3;
+  const unsigned char *p, *d;
+  int i, j, len;
+
+  // ClientHello buf and len
+  d = p  = (unsigned char *)s->init_msg;
+  long n = s->init_num;
+
+  // Get version
+  int version = (((int)p[0]) << 8) | (int)p[1];
+  ja3 += std::to_string(version) + ',';
+  p += 2;
+
+  // Skip client random
+  p += SSL3_RANDOM_SIZE;
+
+  // Skip session id
+  j = *(p++);
+  p += j;
+
+  // No DTLS handling
+
+  // Get cipher suites
+  n2s(p, len);
+  custom_get_ja3_prefixed(2, p, len, ja3);
+  ja3 += ',';
+
+  // Skip compression
+  i = *(p++);
+  p += i;
+
+  // Get extensions
+  uint16_t type;
+  int size;
+  std::string eclist, ecpflist;
+
+  // Skip length blob
+  p += 2;
+  bool first = true;
+  while (p < d + n) {
+    // Each extension blob is comprised of [2bytes] type + [2bytes] size + [size bytes] data
+    n2s(p, type);
+    n2s(p, size);
+
+    // Elliptic curve points
+    if (type == 0x0a) {
+      const unsigned char *sdata = p;
+      n2s(sdata, len);
+      custom_get_ja3_prefixed(2, sdata, len, eclist);
+    }
+    // Elliptic curve point formats
+    else if (type == 0x0b) {
+      const unsigned char *sdata = p;
+      len                        = *(sdata++);
+      custom_get_ja3_prefixed(1, sdata, len, ecpflist);
+    }
+
+    // Update pointer
+    p += size;
+
+    // Update ja3 string with valid extension type
+    if (GREASE_table.find(type) == GREASE_table.end()) {
+      if (!first) {
+        ja3 += '-';
+      }
+      first = false;
+      ja3 += std::to_string(type);
+    }
+  }
+
+  // Append eclist and ecpflist
+  ja3 += "," + eclist + "," + ecpflist;
+  TSDebug(PLUGIN_NAME, "ja3 string: %s", ja3.c_str());
+  return ja3;
+}
+#elif OPENSSL_VERSION_NUMBER >= 0x10101000L
```

Comment:
I'm unclear on why this is expected to work. It seems to select a different parsing of the Client HELLO based on the ATS linked version of openSSL. But how does that control the layout of the Client HELLO, which was constructed by the user agent with some potentially different version of openSSL? Is that some of the parsing can be done with openSSL calls instead of by hand?

Reply:
In OpenSSL1.0.2, you can access the members of SSL structure (including client hello buffer), but no APIs exist. OpenSSL1.1.1 makes the structure opaque (hence no manual parsing) and introduces APIs to access client hello values.

As for the layout of client hello message, I think it is the same across all ssl implementations.